File: BinarySearch/cpp/minimum_in_rotated_sorted_array.cpp
//
// Created by santosh on 12/30/25.
//

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:

    static int getMinimumElementInArotatedSortedArrayWithDuplicates(vector<int> nums)
    {
        int low = 0, high = nums.size()-1;
        while (low<high)
        {
            int mid= low + (high-low)/2;

            if (nums[low] == nums[mid] && nums[mid] == nums[high])
            {
                low++;
                high--;
                continue;
            }

            if (nums[mid] > nums[high])
            {
                low = mid+1;
            } else
            {
                high = mid;
            }
        }

        return nums[low];

    };


    static int getMinimumElementInArotatedSortedArray(vector<int> nums)
    {
        int low = 0, high = nums.size()-1;
        while (low<high)
        {
            int mid= low + (high-low)/2;
            if (nums[mid]>nums[high])
            {
                low = mid+1;
            } else
            {
                high = mid;
            }
        }

        return nums[low];

    };
};

int main()
{
    vector<int> nums = {4,5,1,2,3};
    Solution s;
    cout<<s.getMinimumElementInArotatedSortedArray(nums)<< endl;
}

File: BinarySearch/python/minimum_in_rotated_sorted_array.py
# Minimum in rotated sorted array (LeetCode 153 & 154)
# Created by santosh

class Solution:
    @staticmethod
    def get_minimum_element_in_rotated_sorted_array_with_duplicates(
        nums: list[int],
    ) -> int:
        """
        Handles duplicates by shrinking both ends.
        Time: O(log n) average, O(n) worst case with duplicates
        """
        low, high = 0, len(nums) - 1

        while low < high:
            mid = low + (high - low) // 2

            if nums[low] == nums[mid] == nums[high]:
                low += 1
                high -= 1
                continue

            if nums[mid] > nums[high]:
                low = mid + 1
            else:
                high = mid

        return nums[low]

    @staticmethod
    def get_minimum_element_in_rotated_sorted_array(nums: list[int]) -> int:
        """
        No duplicates version.
        Time: O(log n), Space: O(1)
        """
        low, high = 0, len(nums) - 1

        while low < high:
            mid = low + (high - low) // 2
            if nums[mid] > nums[high]:
                low = mid + 1
            else:
                high = mid

        return nums[low]


if __name__ == "__main__":
    nums = [4, 5, 1, 2, 3]
    s = Solution()
    print(s.get_minimum_element_in_rotated_sorted_array(nums))
